### PR TITLE
Add SANA-WM initial release

### PR DESCRIPTION
**Pull request by HaoyiZhu**
## SANA-WM: initial release (bidirectional inference)

Initial public release of **SANA-WM**, an efficient minute-scale camera-controlled image-to-video diffusion model. This PR adds bidirectional inference only; training / AR variants will follow.

Released checkpoints live at [`Efficient-Large-Model/SANA-WM_bidirectional`](https://huggingface.co/Efficient-Large-Model/SANA-WM_bidirectional) and are auto-downloaded on first use.

### Usage

  ```bash
  bash ./environment_setup_sana_wm.sh sana-wm
  conda activate sana-wm

  python inference_video_scripts/inference_sana_wm.py \
    --image      asset/sana_wm/demo_0.png \
    --prompt     asset/sana_wm/demo_0.txt \
    --action     "w-80,jw-40,w-40,lw-60,w-100" \
    --num_frames 321 \
    --output_dir results/demo
```
Camera trajectory can be either a (F, 4, 4) .npy (--camera) or a WASD/IJKL DSL (--action). Intrinsics are optional — Pi3X-estimated from the input image when omitted. Output is fixed at 704 × 1280 and composited with a WASD + joystick overlay. See `docs/sana_wm.md` for full options.

### What's in

  - Stage-1 2.6 B Sana DiT (hybrid GDN + softmax, dual-branch camera control) + LTX-2 sink-bidirectional refiner.
  - Slim inference config at configs/sana_wm/sana_wm_1600m_720p.yaml.
  - Single-file CLI at inference_video_scripts/inference_sana_wm.py (action string, Pi3X auto-intrinsics, action overlay, --offload_* flags).
  - Three demo scenes under asset/sana_wm/, docs at docs/sana_wm.md, isolated env via environment_setup_sana_wm.sh.

### CODE REVIEW

**Comment by HaozheLiu-ST** [comment]
### TODO 
change the refiner ckpt to diffusers version.

File: configs/sana_wm/sana_wm_1600m_720p.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  cam_attn_compress: 1
+  init_cam_from_base: true
+  use_chunk_plucker_post_attn: true
+  chunk_plucker_channels: 48
+  chunk_plucker_post_attn_blocks: 20
+vae:
+  vae_type: LTX2VAE_diffusers
+  # AutoencoderKLLTX2Video.from_pretrained(<root>, subfolder="vae") expects
+  # a parent directory containing a ``vae/`` diffusers folder. The public
+  # release repo hosts that ``vae/`` folder at its root.
```
Comment (lawrence-cj) [inline configs/sana_wm/sana_wm_1600m_720p.yaml]:
```suggestion
```

Let's remove all useless comments

File: asset/docs/sana_wm.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
+<p align="center" style="border-radius: 10px">
+  <img src="https://raw.githubusercontent.com/NVlabs/Sana/refs/heads/main/asset/sana-wm-logo.png" width="70%" alt="SANA-WM Logo"/>
+</p>
```
Comment (lawrence-cj) [inline asset/docs/sana_wm.md]: The logo is missing.

**Review by yujincheng08 — CHANGES_REQUESTED** [review]
non-chunkwise kernel can be removed

**Comment by Sunwood-ai-labs** [comment]
Thanks for the initial SANA-WM release! I tested PR #379 locally on a Windows 11 + Docker Desktop setup with an RTX 4090 24GB.

What worked:

- Checked out `pull/379/head`
- Built a CUDA Docker image around the PR branch
- Ran bidirectional SANA-WM inference on the provided `asset/sana_wm/demo_0.*` sample
- Successfully generated a short Stage-1/VAE video with:
  - `--num_frames 80`
  - `--step 20`
  - `--cfg_scale 5.0`
  - `--flow_shift 8.0`
  - `--no_refiner --offload_vae`
  - `--intrinsics asset/sana_wm/demo_0_intrinsics.npy`

One thing I ran into: `configs/sana_wm/sana_wm_1600m_720p.yaml` currently sets:

```yaml
vae_pretrained: hf://Efficient-Large-Model/SANA-WM_bidirectional
```

Because `resolve_hf_path()` treats this as a full repo snapshot request, VAE construction started downloading the whole HF repo, including the refiner and Gemma text encoder. On my machine that pulled tens of GB and made it look like VAE loading was hanging.

Changing it locally to the raw HF repo id made the VAE load immediately through Diffusers' `subfolder=vae` path:

```yaml
vae_pretrained: Efficient-Large-Model/SANA-WM_bidirectional
```

After that, `AutoencoderKLLTX2Video.from_pretrained(..., subfolder=vae)` loaded quickly and Stage-1 generation completed.

This may be worth adjusting in the config or documenting, especially for people trying `--no_refiner` or first-time local inference on consumer GPUs. Thanks again for releasing this!

**Comment by Sunwood-ai-labs** [comment]
I also opened a tiny PR directly against the `feat/sana-wm` branch with the one-line config change, in case it is useful to cherry-pick or merge:

https://github.com/HaoyiZhu/NVlabs-Sana/pull/1

File: configs/sana_wm/sana_wm_1600m_720p.yaml (hunk 1 above)
Comment (HaoyiZhu) [reply configs/sana_wm/sana_wm_1600m_720p.yaml]: Done in 48bba61 — removed the 3-line comment.

File: asset/docs/sana_wm.md (hunk 2 above)
Comment (HaoyiZhu) [reply asset/docs/sana_wm.md]: The logo file was being caught by the global `*.png` .gitignore rule. Whitelisted `asset/sana-wm-logo.png`, committed the binary, and switched the two doc refs to relative paths (matches the main `README.md` style). Should render in the diff now — 48bba61.

**Comment by Sunwood-ai-labs** [comment]
Follow-up from my RTX 4090 local testing:

I continued testing this PR on a Windows 11 + Docker Desktop machine with a single RTX 4090 24GB, and was able to get a fuller local workflow running.

Tested branch / changes:

https://github.com/Sunwood-ai-labs/Sana/tree/sana-wm-pr-379-rtx4090-docker

What I verified locally:

- Docker Compose based SANA-WM inference setup
- Stage-1 + VAE decode on RTX 4090
- LTX-2 refiner enabled with `--offload_refiner`
- Sequential CPU offload for the refiner transformer blocks, so the whole refiner transformer is not moved to GPU at once
- Pi3X installed in the Docker image for `--intrinsics`-less runs
- Pi3X-estimated intrinsics from a custom generated input image
- 321-frame request with refiner enabled

The most useful finding: with the PR implementation as-is, the refiner path OOMed on 24GB when moving the full LTX-2 refiner transformer to CUDA. I got it working by keeping the non-block/static transformer modules on GPU and moving `transformer_blocks` to CUDA one block at a time, then moving each block back to CPU and clearing CUDA cache.

I also had to reduce a memory peak in prompt encoding: pack Gemma hidden states on CPU, delete the Gemma text encoder first, then move `prompt_embeds` to GPU for the connectors. This mattered especially after running Pi3X in the same process.

Pi3X notes:

- Added `pip install git+https://github.com/yyfz/Pi3.git --no-deps` to the Docker workflow
- Without extra cleanup, Pi3X + refiner could still hit a later OOM
- Explicitly moving the Pi3X model back to CPU, deleting it, and calling `torch.cuda.ipc_collect()` before loading the heavy SANA/refiner path made the full run pass

Example successful run result:

- Input: custom generated `neon_alley.png`
- Intrinsics: estimated by Pi3X, no `--intrinsics` argument
- Pi3X estimate: `fx=1604.6, fy=1593.3, cx=834.8, cy=469.3`, FOV `H=55.0°, V=32.9°`
- `--num_frames 321`
- saved MP4: 1280x704, 16 fps, 20.000 sec, 320 frames
- `--step 20 --cfg_scale 5.0 --flow_shift 8.0`
- `--offload_vae --offload_refiner`
- total wall time: about 484 sec
- peak RTX 4090 VRAM observed: about 24.0 GB

I also repeated the same seed/settings and got byte-identical output for the same input, which was nice for comparison/debugging.

This is probably not polished enough to merge as-is, but it may be useful as a consumer-GPU reference path or as notes for a future local Docker workflow. Happy to open a focused PR if you would like any of these pieces split out, for example:

- Docker Compose helper for local SANA-WM inference
- `vae_pretrained` config adjustment/documentation
- refiner sequential CPU offload path
- Pi3X Docker dependency + memory cleanup for `--intrinsics`-less inference

Thanks again for releasing the initial SANA-WM implementation.

**Comment by yujincheng08** [comment]
有些工具类的脚本就没必要放了，比如 scan_cam_triton 之类的

**Comment by HaoyiZhu** [comment]
Done in 694f58c — dropped ~7.8k LOC of CP-only and training-utility code that the single-GPU inference release never touches:

- `diffusion/distributed/` (CP setup + halo/scan kernels)
- `diffusion/model/ops/frame_gdn/` (scan_triton, fused_recurrent_triton, api)
- `diffusion/model/ops/fused_gdn_cp.py`
- `diffusion/model/ops/fused_gdn_chunkwise_bwd.py` and `fused_gdn_chunkwise_stateful_bwd.py`
- CP-only forward branches and `triton_*` `update_rule_func` entries in the 4 live net files

Verified: smoke test (`--num_frames 321 --step 60 --no_refiner` with default seed) produces byte-identical mp4 output (md5 `0343e294…`) vs the prior cleanup baseline.

**Comment by johndpope** [comment]
at the risk of coming across spammy - i actually implemented scd on ltx2 - 
https://github.com/johndpope/ltx2-castlehill

training run 
https://wandb.ai/snoozie/scd-finetune/runs/lzod5m2l?nw=nwusersnoozie
this rips out a bunch of layers making arbitary length videos possible on consumer hardware.

im distacted with other stuff (https://www.reddit.com/r/LocalLLaMA/comments/1tee5ms/can_a_5090_with_qwen36_achieve_3000_toks_bring/) - but could use some help here - it would accelerate this project. 
this is the paper - https://arxiv.org/abs/2602.10095

**Comment by lawrence-cj** [comment]
All CI tests are passed. Great team work! @HaoyiZhu @HaozheLiu-ST @yujincheng08 @Owen718 @xieenze @HeliosZhao

**Review by yujincheng08 — APPROVED** [review]
lgtm

**Comment by Sunwood-ai-labs** [comment]
I rebased the RTX 4090 Docker/offload work on top of the latest PR #379 head (`694f58c`) and opened a small companion PR against the `feat/sana-wm` branch here:

https://github.com/HaoyiZhu/NVlabs-Sana/pull/2

It keeps the scope focused on the consumer-GPU/local workflow:

- Docker Compose helper for local SANA-WM inference
- local Dockerfile with Pi3X installed for `--intrinsics`-less runs
- sequential CPU offload for LTX-2 refiner transformer blocks
- Pi3X/Gemma memory cleanup to avoid RTX 4090 24GB OOMs

The earlier `vae_pretrained` path issue is no longer included as a diff because the latest PR branch already has the fixed raw HF repo id.

Local checks after rebasing:

- `docker compose -f docker-compose.sana-wm-pr.yml config`
- `docker compose -f docker-compose.sana-wm-pr.yml build sana-wm-pr`

Both passed locally.

**Comment by lawrence-cj** [comment]
@HaoyiZhu please check this [PR](https://github.com/NVlabs/Sana/pull/379#issuecomment-4504934634), then I think we may merge into our main stream

**Comment by HaoyiZhu** [comment]
> I rebased the RTX 4090 Docker/offload work on top of the latest PR #379 head (`694f58c`) and opened a small companion PR against the `feat/sana-wm` branch here:
> 
> [HaoyiZhu#2](https://github.com/HaoyiZhu/NVlabs-Sana/pull/2)
> 
> It keeps the scope focused on the consumer-GPU/local workflow:
> 
> * Docker Compose helper for local SANA-WM inference
> * local Dockerfile with Pi3X installed for `--intrinsics`-less runs
> * sequential CPU offload for LTX-2 refiner transformer blocks
> * Pi3X/Gemma memory cleanup to avoid RTX 4090 24GB OOMs
> 
> The earlier `vae_pretrained` path issue is no longer included as a diff because the latest PR branch already has the fixed raw HF repo id.
> 
> Local checks after rebasing:
> 
> * `docker compose -f docker-compose.sana-wm-pr.yml config`
> * `docker compose -f docker-compose.sana-wm-pr.yml build sana-wm-pr`
> 
> Both passed locally.

Thanks so much for the detailed write-up and your great effort in optimizing the RTX 4090 workflow! Since the specific implementation doesn't quite align with the current architectural plans for this branch, we may open a new PR in the future to properly integrate your excellent offload and memory cleanup ideas. We really appreciate your valuable contribution!
